Source/Particles/ParticleCreation/SmartCreate.H

```diff
@@ -0,0 +1,120 @@
+/* Copyright 2019-2020 Luca Fedeli, Neil Zaim
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#ifndef SMART_CREATE_H_
+#define SMART_CREATE_H_
+
+#include "DefaultInitialization.H"
+
+#include <AMReX_AmrCore.H>
+#include <AMReX_GpuContainers.H>
+#include <AMReX_ParallelDescriptor.H>
+
+enum SmartCreateMode { default_mode, overwrite_weight };
+
+/**
+ * \brief This is a functor for performing a "smart create" that works
+ * in both host and device code.
+ *
+ * A "smart" create does the following:
+ * First, it initializes the position, cpuid and id of
+ * the particle (in most cases IDs should be initialized using setNewParticleIDs
+ * after their creation).
+ * Then, it initializes all the other components according to initialization policies.
+ * Finally, if SmartCreateMode::overwrite_weight is selected as a template argument, it uses
+ * the weight argument to replace the weight component of the particle.
+ *
+ * You don't create this directly - use the SmartCreateFactory object below.
+ *
+ * @tparam Mode if SmartCreateMode::overwrite_weight is selected, weight argument is used to replace the weight component of the particle
+ */
+
+template<SmartCreateMode Mode>
+struct SmartCreate
+{
+    const InitializationPolicy* m_policy_real;
+    const InitializationPolicy* m_policy_int;
+    const int m_weight_index;
+
+    template <typename PartData>
+    AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+    void operator() (
+        PartData& prt, const int i_prt,
+        const amrex::Real x = 0.0,
+#if (AMREX_SPACEDIM == 3)
+        const amrex::Real y = 0.0,
+#endif
+        const amrex::Real z = 0.0,
+        const amrex::Real weight = 0.0,
+        const int cpu = 0,
+        const int id = 0) const noexcept
+    {
+        prt.m_aos[i_prt].pos(0) = x;
+#if (AMREX_SPACEDIM == 3)
+        prt.m_aos[i_prt].pos(1) = y;
+#endif
+        prt.m_aos[i_prt].pos(2) = z;
+
+        prt.m_aos[i_prt].cpu() = cpu;
+        prt.m_aos[i_prt].id() = id;
+
+         // initialize the real components
+         for (int j = 0; j < PartData::NAR; ++j)
+             prt.m_rdata[j][i_prt] = initializeRealValue(m_policy_real[j]);
+         for (int j = 0; j < prt.m_num_runtime_real; ++j)
+             prt.m_runtime_rdata[j][i_prt] = initializeRealValue(m_policy_real[j+PartData::NAR]);
+
+         // initialize the int components
+         for (int j = 0; j < PartData::NAI; ++j)
+             prt.m_idata[j][i_prt] = initializeIntValue(m_policy_int[j]);
+         for (int j = 0; j < prt.m_num_runtime_int; ++j)
+             prt.m_runtime_idata[j][i_prt] = initializeIntValue(m_policy_int[j+PartData::NAI]);
+
+         //Initialize the weight
```

Comment:
At some point later on, you will need to set other properties of the particle, e.g. momentum, correct? Does it make sense to set the weight at that point, and have this functor just be about initialization?

Reply:
For the Schwinger process the Intialization Policies are enough for all the properties except `weight`: particles come into existence at rest and with an optical depth component extracted from an exponential distribution. This is the motivation behind this design choice.
However, I acknowledge that allowing to modify only the weight is a bit awkward. If you think that it could be preferable, I can remove this option. It would probably make the functor simpler...

Reply:
Are there other hypothetical processes where we might want to SmartCreate and then override some other quantity? In that case, I think it would be simpler to have the general "SmartCreate" just initialize, and then use a second "transform" functor that is specific to the process to set the other quantities.

Reply:
I don't know... but I think that it is wise to foresee the use of a `transform` functor. I will put back the `WIP` label and do the necessary modifications...

Reply:
I've just removed the option to overwrite the `weight`.

Reply:
Thanks!